### PR TITLE
Add zero-shot visual prompting task

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] Update data pipeline to support multiple prompts
- [x] Implement zero-shot learning
    - [x] `learn`
    - [x] `infer`
    - [x] refactoring
    - [x] write tests

TODOs
- Set regression test dataset & regression test

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/core/data/transform_libs/torchvision.py [hunk 1]
```diff
@@ -8,19 +8,51 @@
...
+        if tvt_v2._utils.check_type(  # noqa: SLF001
+            inpt,
+            (
+                F.is_pure_tensor,
+                tv_tensors.Image,
+                PIL.Image.Image,
+                tv_tensors.Video,
+                tv_tensors.Mask,
+                tv_tensors.BoundingBoxes,
+                Points,
```
Comment (vinnamkim) [inline src/otx/core/data/transform_libs/torchvision.py:41]: Is this problem coming from adding `Points`?

File: src/otx/core/model/module/visual_prompting.py [hunk 2]
```diff
@@ -211,3 +237,108 @@ def _inference_step(self, metric: MetricCollection, inputs: VisualPromptingBatch
...
+    def state_dict(self) -> dict[str, Any]:
+        """Return state dictionary of model entity with reference features, masks, and used indices."""
+        state_dict = super().state_dict()
+        state_dict.update(
+            {
+                "model.model.reference_feats": self.model.model.reference_feats,
+                "model.model.reference_masks": self.model.model.reference_masks,
+                "model.model.used_indices": self.model.model.used_indices,
+            },
+        )
```
Comment (vinnamkim) [inline src/otx/core/model/module/visual_prompting.py]: Can we move this to `OTXZeroShotVisualPromptingModel` and `load_state_dict()` too? `OTXModel` should be able to save and load its state dict as standalone.

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py [hunk 3]
```diff
@@ -0,0 +1,697 @@
...
+                "freeze_image_encoder": kwargs.get("freeze_image_encoder", True),
+                "freeze_mask_decoder": kwargs.get("freeze_mask_decoder", True),
+                "freeze_prompt_encoder": kwargs.get("freeze_prompt_encoder", True),
+            },
+        )
+        return kwargs
+
+    def initialize_reference_info(self) -> None:
+        """Initialize reference information."""
+        self.reference_feats: Tensor = None
```
Comment (vinnamkim) [inline src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: I'm feeling that it is more natural to make it as [ParameterDict](https://pytorch.org/docs/stable/generated/torch.nn.ParameterDict.html#torch.nn.ParameterDict) with `requires_grad=False` because we update this values from the given data. And, I guess it makes also comfortable for handling its checkpoint.

File: src/otx/core/config/data.py [hunk 4]
```diff
@@ -81,6 +89,7 @@ class DataModuleConfig:
     test_subset: SubsetConfig
 
     tile_config: TilerConfig
+    vpm_config: VisualPromptingConfig
```
Comment (vinnamkim) [inline src/otx/core/config/data.py]:
```suggestion
    vpm_config: VisualPromptingConfig = VisualPromptingConfig()
```

Comment (vinnamkim) [reply src/otx/core/config/data.py]: Is this working? If so please change this as well.
```python
tile_config: TilerConfig = TilerConfig()
```

File: src/otx/core/data/transform_libs/torchvision.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:41]: An error occurred due to no `Points` type in the type checklist.

Comment (sungchul1) [reply src/otx/core/data/transform_libs/torchvision.py:41]: https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/_utils.py#L177-L198

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 3 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2834/commits/121a03de8e1dbec252e2046e7de5ddb289ca18b8.

File: src/otx/core/config/data.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/core/config/data.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2834/commits/5142fdef8dd643361b6bc9f0de8b6fdbb5886563.

File: src/otx/core/model/module/visual_prompting.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/core/model/module/visual_prompting.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2834/commits/9ba3007e76ade8212ee9f62b04ddceb71a846bc0.

File: src/otx/algo/visual_prompting/backbones/tiny_vit.py [hunk 5]
```diff
@@ -9,7 +9,7 @@
 
 import torch
 import torch.nn.functional as F  # noqa: N812
-from timm.models.layers import DropPath, to_2tuple, trunc_normal_
+from timm.layers import DropPath, to_2tuple, trunc_normal_
```
Comment (wonjuleee) [inline src/otx/algo/visual_prompting/backbones/tiny_vit.py:12]: Could you explain why this updated?

File: src/otx/algo/visual_prompting/segment_anything.py [hunk 6]
```diff
@@ -116,29 +125,100 @@ def load_checkpoint(
                 f"To manually load {load_from}, try to set it to trainer.checkpoint.",
             )
 
-    def forward(
+    def forward(self, mode: str, *args, **kwargs) -> Any:  # noqa: ANN401
+        """Forward method for visual prompting task."""
+        assert mode in ["finetuning", "learn", "infer"]  # noqa: S101
```
Comment (wonjuleee) [inline src/otx/algo/visual_prompting/segment_anything.py:130]: why do we need to have separated finetuning and learn for segment_anything?

Comment (wonjuleee) [reply src/otx/algo/visual_prompting/segment_anything.py:130]: As far as I understood, learn = finetuning, infer = pseudo-labeling wo prompt in segment_anything (w/ point decoder), while learn=ref_feat extraction, infer=pseudo-labeling w/ prompt in zero_shot_segment_anything. Please correct me if I was wrong.

File: src/otx/algo/visual_prompting/backbones/tiny_vit.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/backbones/tiny_vit.py:12]: `timm.models.layers` is deprecated.
https://github.com/huggingface/pytorch-image-models/blob/main/timm/models/layers/__init__.py#L49

File: src/otx/core/model/entity/visual_prompting.py [hunk 7]
```diff
@@ -17,3 +24,35 @@ class OTXVisualPromptingModel(
...
+        if isinstance(destination, dict):
+            # to save reference_info instead of reference_feats only
+            destination.pop(prefix + "model.reference_info.reference_feats")
+            destination.update({prefix + "model.reference_info": self.model.reference_info})
+        return destination
+
+    def load_state_dict_pre_hook(self, state_dict: dict[str, Any], prefix: str = "", *args, **kwargs) -> None:
+        """Load reference info manually."""
+        self.model.reference_info = state_dict.get(prefix + "model.reference_info", self.model.reference_info)
+        state_dict[prefix + "model.reference_info.reference_feats"] = self.model.reference_info["reference_feats"]
```
Comment (vinnamkim) [inline src/otx/core/model/entity/visual_prompting.py:58]: Do we still need this even if we register them as `ParameterDict`? That means the native PyTorch handler for checkpoint can save and load them by itself.

File: src/otx/algo/visual_prompting/segment_anything.py (hunk 6 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/segment_anything.py:130]: `finetuning` uses original SAM forward with batch-wise prompts like other tasks (cls, det, ...).
`learn` also uses original SAM forward, but it uses prompt-wise prompts. `infer` uses PerSAM forward pipeline. So, splitting forward pipeline is required.

File: src/otx/core/model/entity/visual_prompting.py (hunk 7 above)
Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:58]: Because there are some components consisting of not only `Tensor` (`model.reference_info.reference_feats`) but also `list` (`model.reference_info.reference_masks`) and `dict` (`model.reference_info.used_indices`). As you said, native PyTorch handler can save and load `Tensor` automatically, but it seems not to be able to handle `list` and `dict`.

Comment (vinnamkim) [reply src/otx/core/model/entity/visual_prompting.py:58]: I see, but do we need a lot of arguments here? Is `self` sufficient?

Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:58]: `destination` is required to get children modules' state dict and update them to root `destination` at https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L1895.
image

`prefix` is required to apply  lightning's prefix (currently `model.`).
`keep_vars` isn't used in this case, but I added it like the above others.

If you have better idea, please recommend :)

Comment (vinnamkim) [reply src/otx/core/model/entity/visual_prompting.py:58]: Thanks for letting me know this. I didn't have a background on how `nn.Module.state_dict()` working. However, I'm honestly feeling that we need more cleaning up for this. However, let's revisit afterwards.

Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:58]: I agree with that.

**Comment by sungchul1** [comment]
@harimkang zero-shot doesn't need optimizer and scheduler, so I fixed some parts at https://github.com/openvinotoolkit/training_extensions/pull/2834/commits/d6bbfa3134b1d296fee45c07fddfbaa51126fb0a. Could you check if they are correctly fixed?

File: src/otx/engine/engine.py [hunk 8]
```diff
@@ -127,10 +128,10 @@ def __init__(
                 meta_info=self._datamodule.meta_info if self._datamodule is not None else None,
             )
         )
-        self.optimizer: OptimizerCallable = (
+        self.optimizer: OptimizerCallable | None = (
```
Comment (harimkang) [inline src/otx/engine/engine.py:131]: This is typically used for `OTXLitModule` builds. It shouldn't be able to accept `None` at `OTXLitModule`, but it's worth checking. If all is well, it looks fine.

File: src/otx/engine/utils/auto_configurator.py [hunk 9]
```diff
@@ -256,7 +257,7 @@ def get_optimizer(self) -> OptimizerCallable:
         Returns:
             OptimizerCallable: The optimizer callable.
```
Comment (harimkang) [inline src/otx/engine/utils/auto_configurator.py]: It looks good to add `None` as return Type here. (scheduler as well)

File: src/otx/engine/engine.py (hunk 8 above)
Comment (sungchul1) [reply src/otx/engine/engine.py:131]: `optimizer` and `scheduler` can be set in `configure_optimizer`, but I skipped it. So, no problem.
```python
    def configure_optimizers(self) -> None:  # type: ignore[override]
        """Skip configure_optimizers unused in zero-shot visual prompting."""
```
https://github.com/sungchul2/training_extensions/blob/bb0a2c54e6bd03b6d726ef25e344496be1b933fe/src/otx/core/model/module/visual_prompting.py#L268-L269

File: src/otx/engine/utils/auto_configurator.py (hunk 9 above)
Comment (sungchul1) [reply src/otx/engine/utils/auto_configurator.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/2834/commits/dd44ef6ce001cc0715364747f2f7bbab19a0c2af.

**Review by vinnamkim — APPROVED** [review]

**Review by harimkang — APPROVED** [review]

**Review by wonjuleee — APPROVED** [review]
LGTM
